### ISSUE TITLE
Socket leak starting with async-std 1.6.4

### ISSUE THREAD

**Issue by tailhook** [async-rs/async-std#891]
This example:
```rust
async fn connect() {
    loop {
        match async_std::os::unix::net::UnixStream::connect("./non-existent").await {
            Ok(_) => unreachable!(),
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {}
            Err(e) => panic!("Unexpected error: {}", e),
        }
    }
}

fn main() {
    async_std::task::block_on(connect());
}
```
Crashes with `thread 'main' panicked at 'Unexpected error: Too many open files (os error 24)', examples/test-connect.rs:6:23` while it shouldn't and didn't crash in 1.6.3. The crash means file descriptors are leaking (maybe because of a memory leak, but I didn't investigate further).

This file descriptor leak is still there in 1.6.5 and master as of f09fa420.

**Comment by yoshuawuyts** [comment]
Oh no; thanks for the report! This may be related to #888, which was tracked back to the changes in the latest `async-io`. Would you mind filing an issue with the `async-io` crate as well?

**Comment by yoshuawuyts** [comment]
Oh actually; we never updated the `UnixStream` impl. A patch similar to https://github.com/async-rs/async-std/pull/889 may do the trick.

**Comment by tailhook** [comment]
So stjepang/async-io#31 is fixed and this bug is still there on async-io 1.1.5.

**Comment by stjepang** [comment]
The bug was introduced when `socket2` got replaced with `nb-connect`, and the bug was in that `nb-connect` would not auto-close the socket if an error occurred before the end of `connect()`.

I pushed a fix a moment ago in `nb-connect` v1.0.1. If you do `cargo update`, the problem should go away.

Thanks for the bug report!

**Comment by tailhook** [comment]
Yes it works. Thanks for the fix!
